### PR TITLE
[ADVAPP-966]: Introduce the ability for alert statuses to be configurable by institutions

### PR DESCRIPTION
**Pull request by jayushi-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-966

### Technical Description

> Create alert status table and make its relationship with alert table.

### Any deployment steps required?

> No.

### Are any Feature Flags and/or Data Migrations that can eventually be removed Added?

> FeatureFlag: AlertStatusId
Data Migrations:
- app-modules/alert/database/migrations/2024_11_17_062514_data_alert_statuses_table.php
- app-modules/alert/database/migrations/2024_11_17_093824_data_add_data_in_status_id_in_alerts_table.php
- database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php

_______________________________________________

### CODE REVIEW

**Comment by jayushi-canyon** [comment]
@danharrin Solved the test issue. this one ready for review

File: app-modules/student-data-model/src/Filament/Resources/EducatableResource/Pages/Concerns/CanManageEducatableAlerts.php [hunk 1]
```diff
@@ -127,8 +142,12 @@ public function table(Table $table): Table
             ->filters([
                 SelectFilter::make('severity')
                     ->options(AlertSeverity::class),
+                SelectFilter::make('status_id')
+                    ->relationship('status', 'name')
```
Comment (Orrison) [inline app-modules/student-data-model/src/Filament/Resources/EducatableResource/Pages/Concerns/CanManageEducatableAlerts.php:146]: This needs the same `orderBy` as the one above.

File: database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    public function up(): void
+    {
+        $statuses = DB::table('alert_statuses')->select('id', 'classification')->get()->toArray();
+        DB::table('histories')->where('subject_type', 'alert')->eachById(function ($alertHistory, $key) use ($statuses) {
+            $statusId = '';
+            $alertData = json_decode($alertHistory->new);
+
+            foreach ($statuses as $status) {
+                if (isset($alertData->status) && $alertData->status === $status->classification) {
+                    $statusId = $status->id;
```
Comment (Orrison) [inline database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php:50]: You will want to break out of this loop as soon as a match is found
```suggestion
                    $statusId = $status->id;
                    break;
```

File: database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                }
+            }
+
+            unset($alertData->status);
+            $alertData->status_id = $statusId;
+
+            DB::table('histories')->where('id', $alertHistory->id)->update(['new' => json_encode($alertData)]);
```
Comment (Orrison) [inline database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php:57]: You could do all of this above just before the break.

File: database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        });
+    }
+
+    public function down(): void {}
```
Comment (Orrison) [inline database/migrations/2024_11_19_112341_data_change_status_to_status_id_in_histories_table.php:61]: This should have a down which would do that opposite. Find all with matching IDs and changing it to the string of the classification.

File: app-modules/alert/database/migrations/2024_11_14_152015_data_activate_feature_flag_alert_status_id.php
Comment (Orrison) [inline app-modules/alert/database/migrations/2024_11_14_152015_data_activate_feature_flag_alert_status_id.php:1]: This needs to be changed to run as the very last migration of all your migrations.

File: app-modules/alert/src/Models/Alert.php [hunk 5]
```diff
@@ -114,9 +113,9 @@ public function getSubscribable(): ?Subscribable
         return $this->concern instanceof Subscribable ? $this->concern : null;
     }
 
-    public function scopeStatus(Builder $query, AlertStatus $status): void
```
[the remark this answers is not in the document]
Comment (jayushi-canyon) [reply app-modules/alert/src/Models/Alert.php:117]: @Orrison I think I have taken care of this as relationship in all places, So i don't think we would need this scope again now. Let me know what do you think.

**Pull request merged**